### PR TITLE
⚠️ Remove RequeueAfterError

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
`RequeueAfterError` was deprecated in v0.3 by https://github.com/kubernetes-sigs/cluster-api/pull/3387 and follow up PR removed most of its usages.
This PR completes this work by removing a few left-over usages and finally removes `RequeueAfterError`

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3370

### CODE REVIEW

File: controllers/machine_controller.go [hunk 1]
```diff
@@ -533,18 +534,18 @@ func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cl
 
 	if err := kubedrain.RunCordonOrUncordon(ctx, drainer, node, true); err != nil {
 		// Machine will be re-reconciled after a cordon failure.
-		log.Error(err, "Cordon failed")
-		return errors.Errorf("unable to cordon node %s: %v", node.Name, err)
+		log.Error(err, "Cordon failed, retry in 20s")
+		return ctrl.Result{RequeueAfter: 20 * time.Second}, nil
```
Comment (vincepri) [inline controllers/machine_controller.go:538]: Interesting we were not requeueing here before 🤔

File: controllers/machine_controller_phases.go [hunk 2]
```diff
@@ -97,9 +96,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
...
-				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+			// if the reference missing is the infrastructureRef and the machine was up and running, this is an error.
+			if *ref == m.Spec.InfrastructureRef && m.Status.InfrastructureReady {
+				err := errors.Errorf("could not find %v %q for Machine %q in namespace %q, requeuing", ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+				log.Error(err, "Machine infrastructure reference has been deleted after being ready, setting failure state")
+				m.Status.FailureReason = capierrors.MachineStatusErrorPtr(capierrors.InvalidConfigurationMachineError)
+				m.Status.FailureMessage = pointer.StringPtr(fmt.Sprintf("Machine infrastructure resource %v with name %q has been deleted after being ready",
+					m.Spec.InfrastructureRef.GroupVersionKind(), m.Spec.InfrastructureRef.Name))
+				return external.ReconcileOutput{}, err
+			}
```
Comment (vincepri) [inline controllers/machine_controller_phases.go:107]: Why are we changing logic?

File: controllers/machine_controller.go (hunk 1 above)
Comment (fabriziopandini) [reply controllers/machine_controller.go:538]: It felt strange to me, and it was discordant with the existing comment, so I added the requeue.

File: controllers/machine_controller_phases.go (hunk 2 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:107]: We are not changing the logic, we are only moving here a bit of logic from reconcileInfrastructure.

This logic handles a very specific error case, which is considered terminal and triggers a different response that included setting failure reason and message. 
NB. this should be reworked again soon we will move this kind o message to conditions.

File: controllers/machine_controller.go (hunk 1 above)
Comment (detiber) [reply controllers/machine_controller.go:538]: The returned error was causing a re-reconciliation previously, any reason not to continue with the same behavior here now instead of using RequeueAfter here?

File: controllers/machine_controller_phases_test.go [hunk 3]
```diff
@@ -576,12 +576,13 @@ func TestReconcileBootstrap(t *testing.T) {
...
-		name            string
-		bootstrapConfig map[string]interface{}
-		machine         *clusterv1.Machine
-		expectError     bool
-		expected        func(g *WithT, m *clusterv1.Machine)
-		result          *ctrl.Result
+		name               string
+		bootstrapConfig    map[string]interface{}
+		machine            *clusterv1.Machine
+		expectRequeueAfter time.Duration
```
Comment (detiber) [inline controllers/machine_controller_phases_test.go:582]: It doesn't appear that expectRequeueAfter is used anywhere

```suggestion
```

File: controllers/machine_controller_phases.go [hunk 4]
```diff
@@ -192,6 +200,9 @@ func (r *MachineReconciler) reconcileBootstrap(ctx context.Context, cluster *clu
 	if err != nil {
 		return ctrl.Result{}, err
 	}
+	if externalResult.RequeueAfter > 0 {
+		return ctrl.Result{RequeueAfter: externalResult.RequeueAfter}, nil
+	}
```
Comment (detiber) [inline controllers/machine_controller_phases.go:205]: The use of `external.ReconcileOutput` here makes this quite confusing. Not sure I have any good suggestions on how to make it less confusing without requiring more invasive refactoring, though :(

File: controllers/machine_controller_phases.go (hunk 2 above)
Comment (wfernandes) [reply controllers/machine_controller_phases.go:107]: So should we be setting `m.Status.InfrastructureReady` to `false`? Since we are setting the Machine's failure reason and message.

Also we should probably change the `log.Error` and `FailureMessage` to not specifically refer to `Machine Infrastructure resource` since this method is being called by `reconcileBootstrap` as well as `reconcileInfrastructure`.

File: controlplane/kubeadm/controllers/helpers.go [hunk 5]
```diff
@@ -62,33 +61,32 @@ func (r *KubeadmControlPlaneReconciler) reconcileKubeconfig(ctx context.Context,
 			controllerOwnerRef,
 		)
 		if errors.Is(createErr, kubeconfig.ErrDependentCertificateNotFound) {
-			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: dependentCertRequeueAfter},
-				"could not find secret %q, requeuing", secret.ClusterCA)
+			return ctrl.Result{RequeueAfter: dependentCertRequeueAfter}, nil
```
Comment (wfernandes) [inline controlplane/kubeadm/controllers/helpers.go:64]: Do we want to remove the error msg here? It may be useful in KCP line 261 when we log out the error. 

Also, the error should be `secret.Kubeconfig` instead of `secret.ClusterCA`.

File: controlplane/kubeadm/controllers/helpers_test.go [hunk 6]
```diff
@@ -58,7 +59,9 @@ func TestReconcileKubeconfigEmptyAPIEndpoints(t *testing.T) {
 		recorder: record.NewFakeRecorder(32),
 	}
 
-	g.Expect(r.reconcileKubeconfig(ctx, clusterName, clusterv1.APIEndpoint{}, kcp)).To(Succeed())
+	result, err := r.reconcileKubeconfig(ctx, clusterName, clusterv1.APIEndpoint{}, kcp)
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(result).To(Equal(ctrl.Result{}))
```
Comment (wfernandes) [inline controlplane/kubeadm/controllers/helpers_test.go:64]: Since we do want the zero value of `ctrl.Result`
```suggestion
	g.Expect(result).To(BeZero())
```

Comment (wfernandes) [reply controlplane/kubeadm/controllers/helpers_test.go:64]: Same for the occurrences below.

**Review by wfernandes — COMMENTED** [review]
Just some minor nits/clarifications.

**Comment by vincepri** [comment]
Let's address/fix the comments above

**Comment by vincepri** [comment]
Hey folks, what's the status of this PR?

**Comment by vincepri** [comment]
@fabriziopandini Do you have time to rebase this PR?

File: controllers/machine_controller.go (hunk 1 above)
Comment (fabriziopandini) [reply controllers/machine_controller.go:538]: Fixed

File: controllers/machine_controller_phases.go (hunk 2 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:107]: > So should we be setting m.Status.InfrastructureReady to false

No, by setting FailureReason/FailureMessage we are signaling a terminal failure
Also, this would be a behavior change.

> Also we should probably change the log.Error and FailureMessage to not specifically refer to Machine Infrastructure resource

This bit of code is specific for the infrastructure ref only, so the message could be specific

File: controllers/machine_controller_phases_test.go (hunk 3 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases_test.go:582]: removed

File: controlplane/kubeadm/controllers/helpers.go (hunk 5 above)
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/helpers.go:64]: We are replacing a RequeueAfterError with a RequeueAfter, no need for the error anymore

**Comment by fabriziopandini** [comment]
@vincepri this one should be ready to go now

File: controllers/machine_controller.go [hunk 7]
```diff
@@ -537,18 +538,18 @@ func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cl
 
 	if err := kubedrain.RunCordonOrUncordon(ctx, drainer, node, true); err != nil {
 		// Machine will be re-reconciled after a cordon failure.
-		log.Error(err, "Cordon failed")
-		return errors.Errorf("unable to cordon node %s: %v", node.Name, err)
+		log.Error(err, "Cordon failed, retry in 20s")
+		return ctrl.Result{}, errors.Errorf("unable to cordon node %s: %v", node.Name, err)
```
Comment (vincepri) [inline controllers/machine_controller.go:542]: The log line says retry in 20s, but we're returning an error which is going to make the controller go in exponential backoff

File: controllers/machine_controller_phases.go [hunk 8]
```diff
@@ -97,9 +96,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 	obj, err := external.Get(ctx, r.Client, ref, m.Namespace)
 	if err != nil {
 		if apierrors.IsNotFound(errors.Cause(err)) {
-			return external.ReconcileOutput{}, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: externalReadyWait},
-				"could not find %v %q for Machine %q in namespace %q, requeuing",
-				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+			// if the reference missing is the infrastructureRef and the machine was up and running, this is an error.
+			if *ref == m.Spec.InfrastructureRef && m.Status.InfrastructureReady {
+				err := errors.Errorf("could not find %v %q for Machine %q in namespace %q, requeuing", ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
```
Comment (vincepri) [inline controllers/machine_controller_phases.go:101]: This isn't requeue-ing with a fixed timestamp like before?

File: controllers/machine_controller_phases.go [hunk 9]
```diff
@@ -97,9 +96,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
...
+				log.Error(err, "Machine infrastructure reference has been deleted after being ready, setting failure state")
+				m.Status.FailureReason = capierrors.MachineStatusErrorPtr(capierrors.InvalidConfigurationMachineError)
+				m.Status.FailureMessage = pointer.StringPtr(fmt.Sprintf("Machine infrastructure resource %v with name %q has been deleted after being ready",
+					m.Spec.InfrastructureRef.GroupVersionKind(), m.Spec.InfrastructureRef.Name))
+				return external.ReconcileOutput{}, err
+			}
+
+			log.Info(fmt.Sprintf("could not find %v %q for Machine %q in namespace %q, requeueing", ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace))
```
Comment (vincepri) [inline controllers/machine_controller_phases.go:109]:
```suggestion
			log.Infof("Could not find %v %q for Machine %q in namespace %q, requeueing", ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
```

File: controllers/machine_controller_phases.go [hunk 10]
```diff
@@ -97,9 +96,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
...
-				"could not find %v %q for Machine %q in namespace %q, requeuing",
-				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+			// if the reference missing is the infrastructureRef and the machine was up and running, this is an error.
+			if *ref == m.Spec.InfrastructureRef && m.Status.InfrastructureReady {
+				err := errors.Errorf("could not find %v %q for Machine %q in namespace %q, requeuing", ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+				log.Error(err, "Machine infrastructure reference has been deleted after being ready, setting failure state")
+				m.Status.FailureReason = capierrors.MachineStatusErrorPtr(capierrors.InvalidConfigurationMachineError)
+				m.Status.FailureMessage = pointer.StringPtr(fmt.Sprintf("Machine infrastructure resource %v with name %q has been deleted after being ready",
+					m.Spec.InfrastructureRef.GroupVersionKind(), m.Spec.InfrastructureRef.Name))
+				return external.ReconcileOutput{}, err
```
Comment (vincepri) [inline controllers/machine_controller_phases.go:106]: This is new behavior, while this is ok for now, let's try to avoid adding unrelated changes to PRs

File: controllers/machine_controller_phases.go (hunk 2 above)
Comment (vincepri) [inline controllers/machine_controller_phases.go:107]: Let's remove this check, it's something that's not generic for all external resource — it should probably be done in the reconcileInfrastructure bit

**Review by vincepri — CHANGES_REQUESTED** [review]
I'm ~40% way there, will continue review later

File: controllers/machine_controller_phases.go (hunk 9 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:109]: logr.Logger does not support Infof, but keyAndValues 😞

File: controllers/machine_controller_phases.go (hunk 8 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:101]: Reverted to having this logic in reconcileInfrastructure

File: controllers/machine_controller_phases.go (hunk 10 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:106]: Reverted to having this logic in reconcileInfrastructure

File: controllers/machine_controller_phases.go (hunk 2 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:107]: Reverted to having this logic in reconcileInfrastructure

File: controllers/machine_controller_phases.go (hunk 9 above)
Comment (vincepri) [reply controllers/machine_controller_phases.go:109]: let's switch to key values then?

File: controllers/machine_controller.go [hunk 11]
```diff
@@ -537,18 +538,18 @@ func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cl
 
 	if err := kubedrain.RunCordonOrUncordon(ctx, drainer, node, true); err != nil {
 		// Machine will be re-reconciled after a cordon failure.
-		log.Error(err, "Cordon failed")
-		return errors.Errorf("unable to cordon node %s: %v", node.Name, err)
+		log.Error(err, "Cordon failed, retry in 20s")
```
Comment (wfernandes) [inline controllers/machine_controller.go:541]: Should we remove this log or should the following line return `ctrl.Result{RequeueAfter: 20 * time.Second}`?

**Review by wfernandes — COMMENTED** [review]
minor comment.

File: controllers/machine_controller.go (hunk 11 above)
Comment (fabriziopandini) [reply controllers/machine_controller.go]: Removed the log

**Comment by fabriziopandini** [comment]
@vincepri rebased+addressed almost all comments. There is only one point I'm not really happy about...

File: controllers/machine_controller_phases.go [hunk 12]
```diff
@@ -240,14 +241,18 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, cluster
 	// Call generic external reconciler.
 	infraReconcileResult, err := r.reconcileExternal(ctx, cluster, m, &m.Spec.InfrastructureRef)
 	if err != nil {
-		if m.Status.InfrastructureReady && strings.Contains(err.Error(), "could not find") {
-			// Infra object went missing after the machine was up and running
+		return ctrl.Result{}, err
+	}
+	if infraReconcileResult.RequeueAfter > 0 {
```
Comment (fabriziopandini) [inline controllers/machine_controller_phases.go:246]: This is kind of brittle, but TBH also the previous implementation checking the error message wasn't bullet proof...

**Review by vincepri — APPROVED** [review]
/lgtm
/approve
